idaes/apps/grid_integration/pricetaker/clustering.py

```diff
@@ -0,0 +1,249 @@
+
+                    {1: {1: 4, 2: 3, 3: 5},
+                    2: {1: 1, 2: 7, 3: 3}}
+
+
+        weights: dict
+            A dictionary of weights for each representative day, indexed the
+                same way as lmp_data. Example: ::
+
+                    {1: 45, 2: 56}
+    """
+    # reconfiguring raw data
+    daily_data = generate_daily_data(raw_data, horizon_length)
+
+    # KMeans clustering with the optimal number of clusters
+    kmeans = KMeans(n_clusters=n_clusters, random_state=seed).fit(daily_data)
+    centroids = kmeans.cluster_centers_
+    labels = kmeans.labels_
+
+    # Set any centroid values that are < 1e-4 to 0 to avoid noise
+    centroids = centroids * (abs(centroids) >= 1e-4)
+
+    # Create dicts for lmp data and the weight of each cluster
+    # By default, the data is of type numpy.int or numpy.float.
+    # Converting the data to python int/float. Otherwise, Pyomo complains!
+    num_days, num_time_periods = centroids.shape
+    lmp_data_clusters = {
+        d + 1: {t + 1: float(centroids[d, t]) for t in range(num_time_periods)}
+        for d in range(num_days)
+    }
+    weights = {d + 1: int(sum(labels == d)) for d in range(num_days)}
+
+    return lmp_data_clusters, weights
+
+
+def get_optimal_num_clusters(
+    samples: Union[pd.DataFrame, np.array],
+    kmin: int = 2,
+    kmax: int = 30,
+    method: str = "silhouette",
+    generate_elbow_plot: bool = False,
+    seed: int = 42,
+):
+    """
+    Determines the appropriate number of clusters needed for a
+    given price signal.
+
+    Args:
+        samples: pd.DataFrame | np.array,
+            Set of points with rows containing samples and columns
+            containing features
+
+        kmin : int,
+            Minimum number of clusters
+
+        kmax : int,
+            Maximum number of clusters
+
+        method : str,
+            Method for obtaining the optimal number of clusters.
+            Supported methods are elbow and silhouette
+
+        generate_elbow_plot : bool,
+            If True, generates an elbow plot for inertia as a function of
+            number of clusters
+
+        seed : int,
+            Seed value for random number generator
+
+    Returns:
+        n_clusters: int,
+            The optimal number of clusters for the given data
+    """
+    if kmin >= kmax:
+        raise ValueError(f"kmin must be less than kmax, but {kmin} >= {kmax}")
+
+    # For silhouette method, kmin must be 2. So, we require kmin >= 2 in general
+    if kmin <= 1:
+        raise ValueError(f"kmin must be > 1. Received kmin = {kmin}")
+
+    if kmax >= len(samples):
+        raise ValueError(f"kmax must be < len(samples). Received kmax = {kmax}")
+
+    k_values = list(range(kmin, kmax + 1))
+    inertia_values = []
+    mean_silhouette = []
+
+    for k in k_values:
+        kmeans = KMeans(n_clusters=k, random_state=seed).fit(samples)
+        inertia_values.append(kmeans.inertia_)
```

Comment:
The following code has worked better for me. I don't fully understand the underlying mechanism, but the logic is that the current method forces Kmeans to use the same centroid initialization for each k whereas the alternative allows for variation in centroid initialization across different k values. In summary, Kmeans relies on good initial centroids and the alternative reduces the likelihood of repeating a poor initialization
```suggestion
    np.random.seed(seed)
    for k in k_values:
        kmeans = KMeans(n_clusters=k).fit(samples)
        inertia_values.append(kmeans.inertia_)
```